### ISSUE TITLE
Cumulative Sum calculation function

### ISSUE THREAD

**Issue by danger89** [josdejong/mathjs#1633]
Hi,

I saw you got already `sum()` in the libary, but I was missing the [Cumulative Sum](http://mathworld.wolfram.com/CumulativeSum.html).

In JavaScript you could calculate this as follows (using an array as input):
```js
    cumsum(list) {
      let y = 0;
      return list.map(d => y += d);
    }
```

Maybe an idea to add this?

Regards,
Melroy

**Comment by LloydLS** [comment]
Might be nice for the devs to add this if they have time; alternatively, you could use `.map()`:
```
arr = [1,6,3,8,7,10];
cumulativeSum = [];

arr.map( f(x,i,a) = isZero(i[1]) ? cumulativeSum = [x] : cumulativeSum = concat(cumulativeSum, [cumulativeSum[i[1]] + x] ))
// diff == [1, 7, 10, 18, 25, 35] 
```

**Comment by josdejong** [comment]
Adding `cumsum` would be a nice addition Melroy, thanks for your suggestion. Before implementing it we should discuss what to do with 2 or n-dimensional matrices.

Anyone interested in implementing `cumsum`?

**Comment by durgaprasadd** [comment]
I liked the idea of adding functionality of cumulative sum. If you don't mind, I want to implement that `cumsum` . Assign that to me and let me know conventions.

Regards,
Durga Prasad.

**Comment by josdejong** [comment]
Thanks for your offer @durgaprasadd , that would be great. Before you implement the function it is good to agree on what the function will exactly do (what to do with 2 and n-dimensional matrices). Can you do a proposal @durgaprasadd ?

About conventions: the code style is standardjs (`npm run lint`), and you can take an existing function like `sum` as a base to see how things work.

**Comment by hjonasson** [comment]
@josdejong There didn't seem to be movement here so I took this on, giving the same behaviour for n>1 dimensional arrays being calculated in the same way as in numpy. I wanted to ask you some question regarding what I think are details, but would like to do it in the code.

I get denied permission when trying to push to my branch, should I rather fork the repo?

**Comment by josdejong** [comment]
Thanks for picking this up @hjonasson! 

The linter shows a list with errors with the file and line where the issue is, and an explanation which should give a hint on what's wrong. Looking quickly at your commit it looks like you're using 4 space indentation instead of 2.

**Comment by hjonasson** [comment]
I was basically having a hard time configuring my vscode. Using standard can be a bit of a hassle:S I managed this now but you had responded before I edited the comment:D 

I have added most tests now and will push in the morning. The only outstanding issue is that of the two functions I currently have copied from collection.js. If you have pointers for that I can fix it tomorrow most likely. If you don't I can create the PR with the copying in it.

**Comment by hjonasson** [comment]
I figured out the issues I was having and pushed the tests

[PR ready for review](https://github.com/josdejong/mathjs/pull/1765/files)

**Comment by gwhitney** [comment]
Not quite sure exactly how this PR got stalled. I too would like to have a `cumsum` function, as I am extensively working with sequences. And not to complicate matters, but it seems to me that ideally `cumsum([1,2,3])` should be `[0,1,3,6]` so that we have the identity `diff(cumsum(S)) == S` for any sequence S, which would totally make sense. Otherwise (as currently implemented in the PR) we get that `diff(cumsum(S))` is S with the first entry knocked off, which is not as useful. And also if we insert the additional 0 at the beginning of the result of `cumsum` we will get the identity `S == cumsum(diff(S)) + S[0]` which is very nice and intuitive as well. Both of these identities correspond to their continuous counterparts: if F(x) = \int_0^x f(t)dt, then F'(x) = f(x) and if f(x) = F'(x) then F(x) = F(0) + \int_0^x f(t)dt, so for mathematical conformity, I think they are worth having for these discrete diff and cumsum operators.

Of course if cumsum is left without the initial 0, one can get the other variant by prepending a 0 (e.g. `concat([0],cumsum(S))`), and if cumsum is modified to include the initial 0 but some folks don't want it, it could be lopped off as in `cumsum(S)[2:]`. So in some sense it doesn't _really_ matter whether the initial 0 is present; the question is just which is more mathematically natural and/or expected by mathjs users. Obviously, I'd argue that including the 0 is.

Anyhow, please let me know if there is any way I can help get this over the finish line (whichever way you decide on whether the initial 0 should be there).

**Comment by josdejong** [comment]
Thanks for picking this one up @gwhitney.

It's actually a very interesting thing to have `cumsum` be symmetric with `diff`, by adding an initial 0. It makes sense mathematically. Whether this makes sense in practice depends on how `cumsum` will be used typically, so it would be good to collect some typical real world example usages of `cumsum`. Looking at other applications like python and matlab/octave, I see there is no initial 0 returned, and I personally would not expect this either (looking at how I used this for example in Excel in the past). If there is good use cases for both, we could make it an option to add an initial zero or not.

**Comment by gwhitney** [comment]
OK, you're completely right. Every other place I've looked does not include the initial zero that makes diff invert cumsum. I should have done more homework before my posting -- I'm really surprised that nobody else uses the convention that yields these identities. I can totally understand why you wouldn't want mathjs to be the only one that includes the initial 0, even if it is more internally consistent in a way. I will just satisfy myself with first evaluating `xcumsum(S) = concat([0], cumsum(S))` (as long as I am only dealing with 1-dimensional arrays).

OK, all that said, is there anything I can do to help the PR #1765 get over the finish line? I would be very happy to see cumsum get into a mathjs release alongside diff.

**Comment by josdejong** [comment]
Ok let's keep cumsum then without initial zero. It's indeed easy to create wrapper function to add the initial zero when you need it.

I've read through #1765 again: I did take over the PR in #1870, added unit tests there, and a rewrite of a specific compact but hard to understand line of code, and was awaiting feedback from hjonasson but it went stale. So next steps (I think): fix the merge conflicts in #1870, briefly go over it once more, and merge it. And then close #1765 in favor of #1870. Would be great if you would like to pick this up @gwhitney !

**Comment by josdejong** [comment]
The new function `cumsum` is published now in `v10.4.0`
